core/src/Kitty.sol

```diff
@@ -5,183 +5,259 @@ import {ERC20, SafeTransferLib} from "solmate/utils/SafeTransferLib.sol";
-        PackedSlot memory _packedSlot = packedSlot;
+    function accrueInterest() external {
+        (Slot0 memory slot0, Slot1 memory slot1) = _load();
 
         // Reentrancy guard
-        require(_packedSlot.borrowIndexTimestamp != 0);
-        packedSlot.borrowIndexTimestamp = 0;
+        require(slot0.lastAccrualTime != 0);
+        lastAccrualTime = 0;
 
-        uint256 _totalSupply = totalSupply;
-        uint256 _inventory = _getInventory(_packedSlot.totalBorrows);
-        if (_packedSlot.borrowIndexTimestamp == block.timestamp || _inventory == 0) {
-            return (_packedSlot, _totalSupply, _inventory);
-        }
+        (slot0, slot1, ) = _accrueInterest(slot0, slot1);
 
-        uint256 accrualFactor = interestModel.getAccrualFactor(
-            block.timestamp - _packedSlot.borrowIndexTimestamp,
-            FullMath.mulDiv(1e18, _packedSlot.totalBorrows, _inventory)
-        );
-        uint256 accruedInterest = FullMath.mulDiv(_packedSlot.totalBorrows, accrualFactor, 1e18);
+        _save(slot0, slot1); // TODO if inventory == slot0.lastBalance, then no interest was accrued, and we don't need to write slot1 to storage
+    }
+
+    function _accrueInterest(Slot0 memory slot0, Slot1 memory slot1) private returns (Slot0 memory, Slot1 memory, uint256) {
+        (uint256 borrowsOld, uint256 accrualFactor) = _getAccrualFactor(slot0, slot1);
+        if (accrualFactor == 0 || borrowsOld == 0) return (slot0, slot1, slot0.lastBalance); 
 
-        _inventory += accruedInterest;
-        _packedSlot.totalBorrows += uint128(accruedInterest);
-        _packedSlot.borrowIndex = uint96(FullMath.mulDiv(_packedSlot.borrowIndex, 1e18 + accrualFactor, 1e18));
-        _packedSlot.borrowIndexTimestamp = uint32(block.timestamp); // fails after Feb 07 2106 06:28:15
+        // TODO sane constraints on accrualFactor WITH TESTS for when accrualFactor is reported to be massive
+        slot1.borrowIndex = uint72(FullMath.mulDiv(slot1.borrowIndex, INTERNAL_PRECISION + accrualFactor, INTERNAL_PRECISION));
+        slot0.lastAccrualTime = uint32(block.timestamp); // TODO probably don't need to pass this around. removing would make func more pure.
+
+        // re-compute borrows and inventory
+        uint256 borrowsNew = FullMath.mulDiv(slot1.borrowBase, slot1.borrowIndex, BORROWS_SCALER);
+        uint256 inventory;
+        unchecked {
+            inventory = slot0.lastBalance + borrowsNew;
+        }
 
         uint256 newTotalSupply = FullMath.mulDiv(
-            totalSupply,
-            _inventory,
-            _inventory - accruedInterest / 8 // `8` indicates a 12.5% reserve factor
+            slot0.totalSupply,
+            inventory,
+            inventory - (borrowsNew - borrowsOld) / 8 // `8` indicates a 1/8=12.5% reserve factor
         );
-        _mint(treasury, newTotalSupply - totalSupply);
+        if (newTotalSupply != slot0.totalSupply) {
```

Comment:
Context:
`borrowsOld = borrowBase * borrowIndexOld / SCALER`
`borrowsNew = borrowBase * borrowIndexNew / SCALER`

Though `borrowIndexNew` is always greater than `borrowIndexOld`, we're dividing by SCALER to get `borrowsOld` and `borrowsNew`. This decrease in precision implies that `borrowsNew > borrowsOld` may not always be true.

In order for `newTotalSupply != slot0.totalSupply`, not only do we need `borrowsNew > borrowsOld`, but we _also_ need `floor((borrowsNew - borrowsOld) / 8)` to be nonzero. So that delta needs to be at least 8 to change total supply.

In practice, if the Kitty is managing more than ~$100, this will always be true. But on the off chance that a token has weird decimals, or our interest rates are really low, or nobody has borrowed anything yet... checking whether total supply has changed can save us gas.